File: Cpp/AdventOfCode/src/2016/d9_Decompress.cpp
#include "2016/d9_Decompress.h"

SOLUTION(2016, 9) {
    constexpr u64 Decompress(std::string_view line, bool recurse) {
        u64 result = 0;
        u32 runLength;
        u32 repeatCount;
        size_t index = 0;
        while (index < line.size()) {
            if (line[index] != '(') {
                result++;
                index++;
            }
            else {
                auto close = line.find(')', index);
                auto expr = line.substr(index + 1, close - index - 1);
                auto s = Constexpr::Split(expr, "x");
                Constexpr::ParseNumber(s[0], runLength);
                Constexpr::ParseNumber(s[1], repeatCount);
                if (recurse) {
                    result += repeatCount * Decompress(line.substr(close + 1, runLength), recurse);
                }
                else {
                    result += repeatCount * runLength;
                }
                index = close + runLength + 1;
            }
        }

        return result;
    }

    PART_ONE() {
        return Constexpr::ToString(Decompress(Line, false));
    }

    PART_TWO() {
        return Constexpr::ToString(Decompress(Line, true));
    }

    TESTS() {
        if (Decompress("(3x3)XYZ", true) != 9) return false;
        if (Decompress("X(8x2)(3x3)ABCY", true) != 20) return false;
        if (Decompress("(27x12)(20x12)(13x14)(7x10)(1x12)A", true) != 241920) return false;
        if (Decompress("(25x3)(3x3)ABC(2x3)XY(5x2)PQRSTX(18x9)(3x2)TWO(5x7)SEVEN", true) != 445) return false;
        return true;
    }
}

File: Rust/rust_aoc/src/year_2016/d9_decompress.rs
pub fn part_one(lines: &Vec<String>) -> String {
    decompress(&lines[0], false).to_string()
}
pub fn part_two(lines: &Vec<String>) -> String {
    decompress(&lines[0], true).to_string()
}

fn decompress(line: &str, recurse: bool) -> usize {
    let mut result: usize = 0;
    let mut index: usize = 0;

    let bytes = line.as_bytes();
    while index < bytes.len() {
        if bytes[index] != b'(' {
            result += 1;
            index += 1;
        } else {
            let close_distance = line[index..].find(")").unwrap();
            let close_index = index + close_distance;
            let expr = std::str::from_utf8(&bytes[index + 1..close_index]).unwrap();
            let s = expr.split("x").collect::<Vec<&str>>();
            let run_length = s[0].parse::<usize>().unwrap();
            let repeat_count = s[1].parse::<usize>().unwrap();

            let next_expr_index = close_index + 1;
            if recurse {
                result += repeat_count * decompress(std::str::from_utf8(&bytes[next_expr_index..next_expr_index + run_length]).unwrap(), recurse);
            } else {
                result += repeat_count * run_length;
            }
            index = next_expr_index + run_length;
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn part_one_examples() {
        assert_eq!(6, decompress("ADVENT", false));
        assert_eq!(7, decompress("A(1x5)BC", false));
        assert_eq!(9, decompress("(3x3)XYZ", false));
        assert_eq!(11, decompress("A(2x2)BCD(2x2)EFG", false));
        assert_eq!(6, decompress("(6x1)(1x3)", false));
        assert_eq!(18, decompress("X(8x2)(3x3)ABCY", false));
    }

    #[test]
    fn part_two_examples() {
        assert_eq!(241920, decompress("(27x12)(20x12)(13x14)(7x10)(1x12)A", true));
        assert_eq!(445, decompress("(25x3)(3x3)ABC(2x3)XY(5x2)PQRSTX(18x9)(3x2)TWO(5x7)SEVEN", true));
    }
}